### PR TITLE
kubeadm: increase ut coverage for addon/dns

### PR DESCRIPTION
**Pull request by my-git9**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

kubeadm: increase ut coverage for addon/dns

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

File: cmd/kubeadm/app/phases/addons/dns/dns_test.go [hunk 1]
```diff
@@ -899,11 +901,30 @@ func TestEnsureDNSAddon(t *testing.T) {
...
 			},
 			wantOut: "[addons] Applied essential addon: CoreDNS\n",
 		},
+		{
+			name: "get dns replicas failed",
+			args: args{
+				cfg: &kubeadmapi.ClusterConfiguration{
+					DNS: kubeadmapi.DNS{
+						ImageMeta: kubeadmapi.ImageMeta{
+							ImageRepository: "daocloud.io",
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/addons/dns/dns_test.go:915]: let's use a fake image repo - like `foo.bar.io`

File: cmd/kubeadm/app/phases/addons/dns/dns_test.go [hunk 2]
```diff
@@ -1437,9 +1478,182 @@ func TestDeployedDNSAddon(t *testing.T) {
 	}
 }
 
+func TestGetCoreDNSInfo(t *testing.T) {
+	tests := []struct {
+		name          string
+		client        clientset.Interface
+		wantconfigmap *v1.ConfigMap
+		wantcorefile  string
+		wantversion   string
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/addons/dns/dns_test.go:1487]: wantConfigMap
wantCorefile
wantVersion

File: cmd/kubeadm/app/phases/addons/dns/dns_test.go [hunk 3]
```diff
@@ -1437,9 +1478,182 @@ func TestDeployedDNSAddon(t *testing.T) {
...
+		{
+			name:          "no coredns configmap",
+			client:        newMockClientForTest(t, 1, 1, "localhost:4711/coredns/coredns:v1.11.2-pre.1", "", ""),
+			wantconfigmap: nil,
+			wantcorefile:  "",
+			wantversion:   "",
+			wantErr:       false,
+		},
+		{
+			name:          "The key of coredns configmap data does not contain corefile",
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/addons/dns/dns_test.go:1499]:
```suggestion
			name:          "the key of coredns configmap data does not contain corefile",
```

File: cmd/kubeadm/app/phases/addons/dns/dns_test.go [hunk 4]
```diff
@@ -1437,9 +1478,182 @@ func TestDeployedDNSAddon(t *testing.T) {
...
+			client:        newMockClientForTest(t, 1, 1, "localhost:4711/coredns/coredns:v1.11.2-pre.1", "coredns", "Corefilefake"),
+			wantconfigmap: nil,
+			wantcorefile:  "",
+			wantversion:   "",
+			wantErr:       true,
+		},
+		{
+			name:          "Failed to obtain coredns version",
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/addons/dns/dns_test.go:1507]:
```suggestion
			name:          "failed to obtain coredns version",
```

File: cmd/kubeadm/app/phases/addons/dns/dns_test.go [hunk 5]
```diff
@@ -1437,9 +1478,182 @@ func TestDeployedDNSAddon(t *testing.T) {
...
+			client:        newMockClientForTest(t, 1, 2, "localhost:4711/coredns/coredns:v1.11.2-pre.1", "coredns", "Corefile"),
+			wantconfigmap: nil,
+			wantcorefile:  "",
+			wantversion:   "",
+			wantErr:       true,
+		},
+		{
+			name:   "Coredns information can be obtained normally",
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/addons/dns/dns_test.go:1515]:
```suggestion
			name:   "coredns information can be obtained normally",
```

use consistent naming and capitalization

File: cmd/kubeadm/app/phases/addons/dns/dns_test.go [hunk 6]
```diff
@@ -1437,9 +1478,182 @@ func TestDeployedDNSAddon(t *testing.T) {
...
+				t.Errorf("isCoreDNSConfigMapMigrationRequired() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
 // replicas is replica of each DNS deployment
 // deploymentSize is the number of deployments with `k8s-app=kube-dns` label.
-func newMockClientForTest(t *testing.T, replicas int32, deploymentSize int, image string) *clientsetfake.Clientset {
+func newMockClientForTest(t *testing.T, replicas int32, deploymentSize int, image string, configmap string, configdata string) *clientsetfake.Clientset {
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/addons/dns/dns_test.go:1656]:
```suggestion
func newMockClientForTest(t *testing.T, replicas int32, deploymentSize int, image string, configMap string, configData string) *clientsetfake.Clientset {
```

File: cmd/kubeadm/app/phases/addons/dns/dns_test.go [hunk 7]
```diff
@@ -1501,5 +1715,53 @@ func newMockClientForTest(t *testing.T, replicas int32, deploymentSize int, imag
 	if err != nil {
 		t.Fatalf("error creating service: %v", err)
 	}
+
+	if configmap != "" {
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/addons/dns/dns_test.go:1719]:
```suggestion
	if configmap == "" {
		return client
	}
```

**Review by neolit123 — COMMENTED** [review]
thanks
/approve
/triage accepted
/priority backlog

**Review by kinzhi — COMMENTED** [review]
/lgtm

**Pull request merged**